docs/gitops-install.md

```diff
@@ -0,0 +1,144 @@
+- Cluster: The actual software that will be deployed
+
+To properly scope this you'll need an environment name. For the
+purposes of this document we'll call it `my-k3s`.
+
+### Environment Variables
+
+To avoid defining many environment variables we'll simplify by creating an
+`.env` file for our deployment. In this case we'll call it `my-k3s.env` and
+place it where we've cloned understack.
+
+```bash title="/path/to/understack/my-k3s.env"
+UC_REPO="/path/to/understack"
+UC_DEPLOY="/path/to/uc-deploy"
+DEPLOY_NAME="my-k3s"
+```
+
+### Remaining pre-deployment config
+
+ArgoCD will need to know where it can access your deployment config
+repo. This can be over SSH with a key or over HTTPS. Add to your
+`my-k3s.env` the following:
+
+```bash title="/path/to/understack/my-k3s.env"
+UC_DEPLOY_GIT_URL="git@github.com:my-org/uc-deploy.git"
+UC_DEPLOY_SSH_FILE="/path/to/ssh.private"
+```
+
+All services will utilize unique DNS names. The facilitate this, UnderStack
+will take a domain and add sub-domains for them. Add to your `my-k3s.env`
+the following:
+
+```bash title="/path/to/understack/my-k3s.env"
+DNS_ZONE="some.domain.corp"
+```
+
+### Populating the infrastructure
+
+TODO: some examples and documentation on how to build out a cluster
+
+### Generating secrets
+
+Secrets in their very nature are sensitive pieces of data. The ultimate
+storage and injection of these in a production environment needs to be
+carefully considered. For the purposes of this document no specific
+choice has been made but tools like Vault, Sealed Secrets, SOPS, etc
+should be considered. This will only generate the necessary secrets
+using random data to sucessfully continue the installation.
+
+TODO: probably give at least one secure example
+
+```bash
+./scripts/gitops-secrets-gen.sh ./my-k3s.env
+cd /path/to/uc-deploy
+git add secrets/my-k3s
+git commit -m "my-k3s: secrets generation"
+```
+
+### Defining the app deployment
+
+In this section we will use the [App of Apps][app-of-apps] pattern to define
+the deployment of all the components of UnderStack.
+
+```bash
+./scripts/gitops-deploy.sh ./my-k3s.env
+cd /path/to/uc-deploy
+git add clusters/my-k3s
+git commit -m "my-k3s: initial cluster config"
+```
+
+## Final modifications of your deployment
+
+This is point you can make changes to the [ArgoCD][argocd] configs before
+you do the deployment.
+
+## Doing the Deployment
+
+At this point we will use our configs to make the actual deployment.
+Make sure everything you've committed to your deployment repo is pushed
+to your git server so that ArgoCD can access it.
+
+If you do not have ArgoCD deployed then you can use the following:
+
+```bash
+kubectl kustomize --enable-helm \
+    https://github.com/rackerlabs/understack//bootstrap/argocd/ \
+    | kubectl apply -f -
+```
+
+Now configure your ArgoCD to have the credential access to your deploy repo:
+
+```bash
+kubectl -n argocd apply -f /path/to/uc-deploy/secrets/my-k3s/argocd/secret-deploy-repo.yaml
```

Comment:
The indentation in that file is broken:

```
❯ k -n argocd apply -f /home/skrobul/devel/test-uc-deploy/secrets/marek-testrun/argocd/secret-deploy-repo.yaml
error: error parsing /home/skrobul/devel/test-uc-deploy/secrets/marek-testrun/argocd/secret-deploy-repo.yaml: error converting YAML to JSON: yaml: line 10: could not find expected ':'
```

and the file:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: marek-testrun-repo
  labels:
    argocd.argoproj.io/secret-type: repo-creds
data:
  sshPrivateKey: LS0tLS1CRREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDXXX
REDACTEDREDACTEDREDACTEDREALLYREALLYREDACTED313370x0deadb33fMzI0NzkyMgoKMi9y
REDACTEDREDACTEDREDACTEDREALLYREALLYREDACTED313370x0deadb33fMzI0NzkyMgoKMi9y
[...]
  type: Z2l0
  url: 2l0QGdpdREDACTEDREDACTED2J1bC90ZXN0LXVjLWRlcGxveS5naXQ=
```

while the `sshPrivateKey` should be just a single base64 encoded, very long line

Reply:
I believe I've fixed this.